<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>syntactic • syntactic</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="syntactic">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">syntactic</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.6</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/syntactic.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/acidgenomics/syntactic">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>syntactic</h1>
                        <h4 class="author">Michael Steinbaugh</h4>
            
            <h4 class="date">27 January 2020</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/acidgenomics/syntactic/blob/master/vignettes/syntactic.Rmd"><code>vignettes/syntactic.Rmd</code></a></small>
      <div class="hidden name"><code>syntactic.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(syntactic)</a>
<a class="sourceLine" id="cb1-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(syntactic, <span class="dt">package =</span> <span class="st">"acidtest"</span>)</a>
<a class="sourceLine" id="cb1-3" data-line-number="3">object &lt;-<span class="st"> </span>syntactic<span class="op">$</span>character</a></code></pre></div>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>The <a href="https://syntactic.acidgenomics.com/">syntactic</a> package returns syntactically valid names from user-defined sample and other biological metadata. The package improves upon the <code>make.names</code> function defined in base R, specifically by adding smart handling of mixed case acronyms (e.g. mRNA, RNAi), decimals, and other coventions commonly used in the life sciences.</p>
<p>The package is intended to work in two modes: string mode (default) and file rename mode.</p>
<p>There are five primary naming functions:</p>
<ul>
<li>
<code>camelCase</code> (e.g. “helloWorld”).</li>
<li>
<code>dottedCase</code> (e.g. “hello.world”).</li>
<li>
<code>snakeCase</code> (e.g. “hello_world”).</li>
<li>
<code>kebabCase</code> (e.g. “hello-world”).</li>
<li>
<code>upperCamelCase</code> (e.g. “HelloWorld”).</li>
</ul>
</div>
<div id="recommended-naming-conventions" class="section level1">
<h1 class="hasAnchor">
<a href="#recommended-naming-conventions" class="anchor"></a>Recommended naming conventions</h1>
<p>Unsure how to name variables and/or functions in R? Here are my current recommendations, <em>for scientists and bioinformaticians</em>:</p>
<ul>
<li>When in doubt, refer to the <a href="https://bioconductor.org/developers/how-to/coding-style/">Bioconductor coding style guide</a>.</li>
<li>Use <code>snakeCase</code> as your daily driver, inside of scripts and R Markdown files. This convention is always legible and consistent. Don’t use this convention inside of packages, however.</li>
<li>For packages, switch to <code>camelCase</code>. Use this consistently for function names, arguments, and internal variable names. Camel case is the preferred convention for <a href="http://bioconductor.org/">Bioconductor</a>, which uses the S4 class system for object-oriented programming. S4 generics on Bioconductor are primarily named in camel case; refer to <a href="https://bioconductor.org/packages/BiocGenerics/">BiocGenerics</a> for details. Note that snake case is used for function names in <a href="https://rstudio.com/">RStudio</a> / <a href="https://www.tidyverse.org/">tidyverse</a> packages, but these build on top of the S3 class system, which isn’t used inside of most biological R packages.</li>
<li>
<code>upperCamelCase</code> should only ever be used for S4 class definitions, such as <code>SummarizedExperiment</code>. Avoid naming functions with this convention.</li>
<li>Use <code>kebabCase</code> for file names. Dashes (hyphens) serve as consistent word boundaries across platforms, whereas underscores do not. This applies in particular to URLs. <a href="https://blog.codinghorror.com/of-spaces-underscores-and-dashes/">This post</a> by Jeff Atwood explains nicely why you should use dashes instead of underscores for file names.</li>
<li>Avoid using <code>dottedCase</code> in R whenever possible. It’s the original naming convention defined in R, but it’s smart to instead use <code>snakeCase</code> and its convention of underscores instead. The S3 class system uses a naming convention of <code>generic.method</code>, which can get mixed up by variables containing periods (dots) in the name.</li>
<li>Valid names in R can’t start with a number. This is often an issue when importing sequencing data (e.g. FASTQ files). The naming functions will add an “x” prefix in this case.</li>
</ul>
</div>
<div id="string-mode" class="section level1">
<h1 class="hasAnchor">
<a href="#string-mode" class="anchor"></a>String mode</h1>
<p>In general, stick with <code>snakeCase</code> or <code>camelCase</code> when sanitizing character strings in R.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(object)</a></code></pre></div>
<pre><code>##  [1] "%GC"             "10uM"            "5'-3' bias"      "5prime"         
##  [5] "G2M.Score"       "hello world"     "HELLO WORLD"     "Mazda RX4"      
##  [9] "nCount"          "RNAi clones"     "tx2gene"         "TX2GeneID"      
## [13] "worfdbHTMLRemap" "123"</code></pre>
<p>Use snake case formatting inside of scripts.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="kw"><a href="../reference/snakeCase.html">snakeCase</a></span>(object)</a></code></pre></div>
<pre><code>##  [1] "percent_gc"        "x10um"             "x5_3_bias"        
##  [4] "x5prime"           "g2m_score"         "hello_world"      
##  [7] "hello_world"       "mazda_rx4"         "n_count"          
## [10] "rnai_clones"       "tx2gene"           "tx2_gene_id"      
## [13] "worfdb_html_remap" "x123"</code></pre>
<p>We recommend using camel case inside of packages. The syntactic package offers two variants: relaxed (default) or strict mode. We prefer relaxed mode for function names, which generally returns acronyms (e.g. ID) more legibly.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw"><a href="../reference/camelCase.html">camelCase</a></span>(object, <span class="dt">strict =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<pre><code>##  [1] "percentGC"       "x10um"           "x5x3Bias"        "x5prime"        
##  [5] "g2mScore"        "helloWorld"      "helloWORLD"      "mazdaRX4"       
##  [9] "nCount"          "rnaiClones"      "tx2gene"         "tx2GeneID"      
## [13] "worfdbHTMLRemap" "x123"</code></pre>
<p>If you’re more old school and prefer using strict camel conventions, that’s also an option.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="../reference/camelCase.html">camelCase</a></span>(object, <span class="dt">strict =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>##  [1] "percentGc"       "x10um"           "x5x3Bias"        "x5prime"        
##  [5] "g2mScore"        "helloWorld"      "helloWorld"      "mazdaRx4"       
##  [9] "nCount"          "rnaiClones"      "tx2gene"         "tx2GeneId"      
## [13] "worfdbHtmlRemap" "x123"</code></pre>
<p>Here’s the default convention in R, for comparison:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/make.names.html">make.names</a></span>(object)</a></code></pre></div>
<pre><code>##  [1] "X.GC"            "X10uM"           "X5..3..bias"     "X5prime"        
##  [5] "G2M.Score"       "hello.world"     "HELLO.WORLD"     "Mazda.RX4"      
##  [9] "nCount"          "RNAi.clones"     "tx2gene"         "TX2GeneID"      
## [13] "worfdbHTMLRemap" "X123"</code></pre>
<p>Additionally, the package exports these string functions:</p>
<ul>
<li>
<code><a href="../reference/capitalize.html">capitalize()</a></code>: Capitalize the first letter of all words in a string.</li>
<li>
<code>sentenceCase</code>: Convert a string into sentence case.</li>
<li>
<code><a href="../reference/makeNames.html">makeNames()</a></code>: A modern variant of <code><a href="https://rdrr.io/r/base/make.names.html">make.names()</a></code> that sanitizes using underscores instead of dots.</li>
</ul>
</div>
<div id="file-rename-mode" class="section level1">
<h1 class="hasAnchor">
<a href="#file-rename-mode" class="anchor"></a>File rename mode</h1>
<p>The package also supports file name sanitization, using the <code>rename = TRUE</code> argument for supported functions. This currently includes <code>kebabCase</code> (recommended), <code>snakeCase</code>, and <code>camelCase</code>.</p>
<p>Here’s an example of how to quickly rename files on disk into kebab case:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1">input &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(</a>
<a class="sourceLine" id="cb12-2" data-line-number="2">    <span class="st">"mRNA Extraction.pdf"</span>,</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">    <span class="st">"inDrops v3 Library Prep.pdf"</span></a>
<a class="sourceLine" id="cb12-4" data-line-number="4">)</a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.create</a></span>(input))</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">output &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kebabCase.html">kebabCase</a></span>(input, <span class="dt">rename =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb12-7" data-line-number="7"><span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(output)</a></code></pre></div>
<pre><code>## [1] "mrna-extraction.pdf"         "indrops-v3-library-prep.pdf"</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.remove</a></span>(output))</a></code></pre></div>
<p>File names containing a prefix that is considered illegal in R can be allowed, which is often useful for sequencing data:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1">input &lt;-<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">4</span>), <span class="st">"_sample_"</span>, LETTERS[<span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="dv">4</span>)], <span class="st">".fastq.gz"</span>)</a>
<a class="sourceLine" id="cb15-2" data-line-number="2"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(input)</a></code></pre></div>
<pre><code>## [1] "1_sample_A.fastq.gz" "2_sample_B.fastq.gz" "3_sample_C.fastq.gz"
## [4] "4_sample_D.fastq.gz"</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.create</a></span>(input))</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">output &lt;-<span class="st"> </span><span class="kw"><a href="../reference/kebabCase.html">kebabCase</a></span>(input, <span class="dt">rename =</span> <span class="ot">TRUE</span>, <span class="dt">prefix =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb17-3" data-line-number="3"><span class="kw"><a href="https://rdrr.io/r/base/basename.html">basename</a></span>(output)</a></code></pre></div>
<pre><code>## [1] "1-sample-a.fastq.gz" "2-sample-b.fastq.gz" "3-sample-c.fastq.gz"
## [4] "4-sample-d.fastq.gz"</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="https://rdrr.io/r/base/invisible.html">invisible</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/files.html">file.remove</a></span>(output))</a></code></pre></div>
<p>Recursion inside of directories is supported using the <code>recursive = TRUE</code> argument.</p>
<p>Our <a href="https://koopa.acidgenomics.com/">koopa</a> shell bootloader uses these functions internally for quick interactive file renaming. In that package, refer to <code>kebab-case</code>, <code>snake-case</code>, and/or <code>camel-case</code> documentation for details.</p>
</div>
<div id="additional-methods" class="section level1">
<h1 class="hasAnchor">
<a href="#additional-methods" class="anchor"></a>Additional methods</h1>
<p>The syntactic package only contains S4 methods defined for <code>character</code> vectors, to keep the package lightweight with few dependencies. Additional S4 methods for Bioconductor classes, including <code>DataFrame</code>, <code>GenomicRanges</code>, and <code>SummarizedExperiment</code>, are defined in the <a href="https://basejump.acidgenomics.com/">basejump</a> package.</p>
</div>
<div id="related-packages" class="section level1">
<h1 class="hasAnchor">
<a href="#related-packages" class="anchor"></a>Related packages</h1>
<p>If syntactic doesn’t work quite right for your workflow, these popular packages also provide excellent sanitization support:</p>
<ul>
<li>
<a href="https://cran.r-project.org/package=janitor">janitor</a> by Sam Firke.</li>
<li>
<a href="https://cran.r-project.org/package=lettercase">lettercase</a> by Christopher Brown.</li>
<li>
<a href="https://cran.r-project.org/package=snakecase">snakecase</a> by Malte Grosser.</li>
</ul>
</div>
<div id="r-session-information" class="section level1">
<h1 class="hasAnchor">
<a href="#r-session-information" class="anchor"></a>R session information</h1>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">utils<span class="op">::</span><span class="kw"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span>()</a></code></pre></div>
<pre><code>## R version 3.6.2 (2019-12-12)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS Mojave 10.14.6
## 
## Matrix products: default
## BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] syntactic_0.3.6  BiocStyle_2.14.4
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.3          rstudioapi_0.10     knitr_1.27         
##  [4] magrittr_1.5        MASS_7.3-51.5       goalie_0.4.1       
##  [7] acidbase_0.1.4      R6_2.4.1            rlang_0.4.3        
## [10] stringr_1.4.0       tools_3.6.2         xfun_0.12          
## [13] acidgenerics_0.3.1  htmltools_0.4.0     yaml_2.2.0         
## [16] assertthat_0.2.1    digest_0.6.23       rprojroot_1.3-2    
## [19] pkgdown_1.4.1       crayon_1.3.4        bookdown_0.17      
## [22] BiocManager_1.30.10 fs_1.3.1            memoise_1.1.0      
## [25] evaluate_0.14       rmarkdown_2.1       stringi_1.4.5      
## [28] compiler_3.6.2      desc_1.2.0          backports_1.1.5</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#recommended-naming-conventions">Recommended naming conventions</a></li>
      <li><a href="#string-mode">String mode</a></li>
      <li><a href="#file-rename-mode">File rename mode</a></li>
      <li><a href="#additional-methods">Additional methods</a></li>
      <li><a href="#related-packages">Related packages</a></li>
      <li><a href="#r-session-information">R session information</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://acidgenomics.com/">Acid Genomics</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
